src/middleware/err_handlers.rs

```diff
@@ -114,35 +119,67 @@ where
 {
     type Response = ServiceResponse<B>;
     type Error = Error;
-    type Future = LocalBoxFuture<'static, Result<Self::Response, Self::Error>>;
+    type Future = ErrorHandlersFuture<S::Future, B>;
 
     actix_service::forward_ready!(service);
 
     fn call(&mut self, req: ServiceRequest) -> Self::Future {
         let handlers = self.handlers.clone();
         let fut = self.service.call(req);
+        ErrorHandlersFuture::ServiceFuture { fut, handlers }
+    }
+}
 
-        async move {
-            let res = fut.await?;
+#[pin_project::pin_project(project = ErrorHandlersProj)]
+pub enum ErrorHandlersFuture<Fut, B>
+where
+    Fut: Future,
+{
+    ServiceFuture {
+        #[pin]
+        fut: Fut,
+        handlers: Handlers<B>,
+    },
+    HandlerFuture {
+        fut: LocalBoxFuture<'static, Fut::Output>,
+    },
+}
 
-            if let Some(handler) = handlers.get(&res.status()) {
-                match handler(res) {
-                    Ok(ErrorHandlerResponse::Response(res)) => Ok(res),
-                    Ok(ErrorHandlerResponse::Future(fut)) => fut.await,
-                    Err(e) => Err(e),
+impl<Fut, B> Future for ErrorHandlersFuture<Fut, B>
+where
+    Fut: Future<Output = Result<ServiceResponse<B>, Error>>,
+{
+    type Output = Fut::Output;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        loop {
+            match self.as_mut().project() {
+                ErrorHandlersProj::ServiceFuture { fut, handlers } => {
+                    let res = ready!(fut.poll(cx))?;
+                    match handlers.get(&res.status()) {
+                        Some(handler) => match handler(res)? {
+                            ErrorHandlerResponse::Response(res) => {
+                                return Poll::Ready(Ok(res))
+                            }
+                            ErrorHandlerResponse::Future(fut) => self
+                                .as_mut()
+                                .set(ErrorHandlersFuture::HandlerFuture { fut }),
```

Comment:
from a quick skim seems like this is the only branch that breaks out if the loop and would be a bit clearer if it could be unwrapped from the loop and call self.poll() instead

Reply:
I feel loop is always clearer than recursive call but it could be just me.